tomcat/src/main/java/org/apache/coyote/http11/Http11Processor.java

```diff
@@ -29,19 +40,89 @@ public void process(final Socket connection) {
+        String line;
+        while ((line = inputReader.readLine()) != null && !line.isBlank()) {
+            rawHttpRequest.add(line);
+        }
+        if (rawHttpRequest.isEmpty()) {
+            throw new IllegalArgumentException("잘못된 요청입니다.");
+        }
+        return RequestData.of(rawHttpRequest);
+    }
+
+    private String getResponse(RequestData requestData) {
+        try {
+            if (isRootPath(requestData)) {
+                byte[] body = "Hello world!".getBytes(StandardCharsets.UTF_8);
+                return buildSuccessResponse(requestData.getHttpVersion(), HttpContentType.HTML, body);
+            }
+            final String staticFilePath = getStaticFilePath(requestData);
+            final byte[] body = readFile(staticFilePath);
+            final var contentType = HttpContentType.getByFileName(staticFilePath);
+            return buildSuccessResponse(requestData.getHttpVersion(), contentType, body);
+        } catch (UncheckedServletException e) {
+            log.error(e.getMessage(), e);
+        }
+        throw new IllegalArgumentException("응답 생성 중에 오류가 발생했습니다.");
+    }
+
+    private boolean isRootPath(RequestData requestData) {
+        return requestData.getResource() == null
+                || requestData.getResource().isBlank()
+                || requestData.getResource().equals("/");
+    }
+
+    private String buildSuccessResponse(HttpVersion httpVersion, HttpContentType contentType, byte[] body) {
+        return buildResponse(httpVersion, HttpStatus.OK, contentType, body);
+    }
+
+    private String buildResponse(HttpVersion httpVersion, HttpStatus httpStatus, HttpContentType contentType,
+                                 byte[] body) {
+        final var headers = String.join("\r\n",
+                httpVersion.getResponseHeader() + " " + httpStatus.getResponseLabel() + " ",
+                contentType.getResponseHeader() + " ",
+                "Content-Length: " + body.length + " ",
+                "");
+        final var bodyString = new String(body, StandardCharsets.UTF_8);
+        return headers + "\r\n" + bodyString;
+    }
+
+    private String getStaticFilePath(RequestData requestData) {
+        String resource = requestData.getResource();
+        String staticFilePath = "static" + resource;
+        if (requestData.getHttpContentType() == HttpContentType.HTML && !staticFilePath.endsWith(".html")) {
+            staticFilePath += ".html";
+        }
+        return staticFilePath;
+    }
+
+    private byte[] readFile(String staticFilePath) {
+        try {
+            final URL resourceUrl = getClass().getClassLoader().getResource(staticFilePath);
+            if (resourceUrl == null) {
+                throw new IllegalArgumentException("존재하지 않는 리소스입니다.: " + staticFilePath);
+            }
+            String path = resourceUrl.getPath();
+            return Files.readString(Path.of(path)).getBytes(StandardCharsets.UTF_8);
```

Comment:
문자 인코딩 문제가 발생합니다.
윈도우의 파일 경로는 `C:/workspace/woowa` 와 같이 `:` 문자가 포함됩니다.
`공백`, `:` 문자는 URL 인코딩시 `%20`, `%3A` 으로 사용됩니다.

문제가 발생하는 코드는 다음 라인입니다.
```
String path = resourceUrl.getPath();
```
resourceUrl.getPath(); 메서드는 Path 객체가 아닌 URL 인코딩된 String을 응답합니다.
`C:/workspace/woowa`  과 같은 경로는 실제로 다음 문자열이 응답됩니다.
`C%3A/workspace/woowa` 

윈도우 환경이 외에도 경로에 공백이 들어가면 %20으로 인코딩 된 문자열을 응답하기에 문제가 발생합니다.

---

이를 해결하는 방법은 여러가지가 있습니다.
1. URL.getPath() 응답값 String을 URL Decode 이후 Path에 사용.
2. Path.of(staticFilePath) 처럼 바로 Path 객체 생성

저는 2번 방법을 추천드립니다.

```suggestion
            final URL resourceUrl = getClass().getClassLoader().getResource(staticFilePath);
            if (resourceUrl == null) {
                throw new IllegalArgumentException("존재하지 않는 리소스입니다.: " + staticFilePath);
            }
            
            final Path path = Path.of(staticFilePath);
            final File file = new File(resourceUrl.getFile());
            return Files.readAllBytes(file.toPath());
```

Reply:
추가로 Files, File, Path는 jar 빌드 후 문제가 발생할 수 있으니 관심있으면 추가로 학습해보세요!

Reply:
왁 감사합니다. path 만 사용하면 운영체제 문제가 다 해결되는 줄 알았는데..아니였네요ㅠㅠ

getResourceAsStream을 사용하면, 파일 시스템, Jar 여부 상관없이 통일된 접근으로 읽을 수 있다고 해서 이 방식으로 리팩토링했습니다.